apps/lib/app_x509.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include "apps.h"
+
+/*
+ * X509_ctrl_str() is sorely lacking in libcrypto, but is still needed to
+ * allow the application to process verification options in a manner similar
+ * to signature or other options that pass through EVP_PKEY_CTX_ctrl_str(),
+ * for uniformity.
+ *
+ * As soon as more stuff is added, the code will need serious rework.  For
+ * the moment, it only handles the FIPS 196 / SM2 distinguishing ID.
+ */
+static ASN1_OCTET_STRING *mk_octet_string(void *value, size_t value_len)
+{
+    ASN1_OCTET_STRING *v = ASN1_OCTET_STRING_new();
+
+    if (v == NULL) {
+        BIO_printf(bio_err,
+                   "error: distinguishing ID allocation failed\n");
+    } else if (!ASN1_OCTET_STRING_set(v, value, value_len)) {
+        BIO_printf(bio_err,
+                   "error: setting distinguishing ID failed\n");
+        ASN1_OCTET_STRING_free(v);
+        v = NULL;
+    }
+    return v;
+}
+
+static int set_x509_distid(void *object, void *value, size_t value_len)
+{
+    X509 *x = object;
+    ASN1_OCTET_STRING *v = mk_octet_string(value, value_len);
+
+    if (v == NULL)
+        return 0;
+
+    X509_set0_distinguishing_id(x, v);
+    return 1;
+}
+
+static int set_x509_req_distid(void *object, void *value, size_t value_len)
+{
+    X509_REQ *x = object;
+    ASN1_OCTET_STRING *v = mk_octet_string(value, value_len);
+
+    if (v == NULL)
+        return 0;
+
+    X509_REQ_set0_distinguishing_id(x, v);
+    return 1;
+}
+
+static int do_x509_ctrl_string(int (*fn)(void *object,
+                                         void *value, size_t value_n),
+                               void *object, const char *value)
+{
+    int rv = 0;
+    char *stmp, *vtmp = NULL;
+    size_t vtmp_len = 0;
+    int cmd = 0; /* Will get command values that make sense somehow */
+
+    stmp = OPENSSL_strdup(value);
+    if (stmp == NULL)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+        vtmp_len = strlen(vtmp);
+    }
+
+    if (strcmp(stmp, "distid") == 0
+               /* GmSSL compatibility */
```

Comment:
For this comment and the following one, GmSSL is another project's name (https://github.com/guanzhi/GmSSL), and I don't think option `sm2_id` is compatible with GmSSL so this comment is confusing.

Reply:
Could be changed to some like `/* OpenSSL 1.1.1 compatibility or so */`

Reply:
Hmmm, I noticed that these sm2 options were added just after the 1.1.1 release, so it seems we don't need to consider backward compatibility at all

Reply:
Nope. These string controls do not exist at all in 1.1.1. I checked. But you're right, gmssl don't use that, they use "signer_id" (which isn't a bad idea). 

Sooo, is "sm2_id" actually used by anyone other than fans of our master? If not, I'm considering simply removing it.

Reply:
> Sooo, is "sm2_id" actually used by anyone other than fans of our master? If not, I'm considering simply removing it.

I think it's removable

Reply:
Done